File: src/qfn2l/stats.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Statistics tracking for the solver."""

import time
import typing


class Stat:
    """A single trackable statistic that behaves like a number."""

    def __init__(self, name: str, initial: typing.Any = 0, fmt: str = ""):
        self.name = name
        self.value = initial
        self.initial = initial
        self.fmt = fmt

    def reset(self):
        self.value = self.initial

    def __iadd__(self, other):
        self.value += other
        return self

    def __str__(self):
        return f"{self.value:{self.fmt}}" if self.fmt else str(self.value)

    def __repr__(self):
        return str(self.value)

    def __gt__(self, other):
        return self.value > other

    def __ge__(self, other):
        return self.value >= other

    def __lt__(self, other):
        return self.value < other

    def __le__(self, other):
        return self.value <= other

    def __eq__(self, other):
        return self.value == other


class Stats:
    """Statistics tracking for the solver."""

    def __init__(self) -> None:
        self._phase_stack: list[tuple["Stat", float]] = []
        self.its = Stat("iteration", 0)
        self.pures = Stat("pures", 0)
        self.mul_axioms = Stat("mul_axioms", 0)
        self.mod_axioms = Stat("mod_axioms", 0)
        self.div_axioms = Stat("div_axioms", 0)
        self.liacalls = Stat("lia_calls", 0)
        self.liatime = Stat("lia_time", 0.0, ".2f")
        self.parse_time = Stat("parse_time", 0.0, ".2f")
        self.nnf_time = Stat("nnf_time", 0.0, ".2f")
        self.simplify_time = Stat("simplify_time", 0.0, ".2f")
        self.propagate_time = Stat("propagate_time", 0.0, ".2f")
        self.makedefs_time = Stat("makedefs_time", 0.0, ".2f")
        self.init_time = Stat("init_time", 0.0, ".2f")
        self.set_level_time = Stat("set_level_time", 0.0, ".2f")
        self.check_nia_time = Stat("check_nia_time", 0.0, ".2f")
        self.solve_time = Stat("solve_time", 0.0, ".2f")
        self.complete_model_time = Stat("complete_model_time", 0.0, ".2f")
        self.total_time = Stat("total_time", 0.0, ".2f")
        self._all = [
            self.its,
            self.pures,
            self.mul_axioms,
            self.mod_axioms,
            self.div_axioms,
            self.liacalls,
            self.liatime,
            self.parse_time,
            self.nnf_time,
            self.simplify_time,
            self.propagate_time,
            self.makedefs_time,
            self.init_time,
            self.set_level_time,
            self.check_nia_time,
            self.solve_time,
            self.complete_model_time,
            self.total_time,
        ]
        self._containers = [self.init_time, self.solve_time, self.total_time]

    def reset(self) -> None:
        for stat in self._all:
            stat.reset()

    def begin_phase(self, stat: "Stat") -> None:
        self._phase_stack.append((stat, time.perf_counter()))

    def end_phase(self) -> None:
        stat, start = self._phase_stack.pop()
        stat += time.perf_counter() - start

    def commit_phases(self) -> None:
        """Commit all in-progress phases; called from the alarm handler."""
        now = time.perf_counter()
        for stat, start in self._phase_stack:
            stat += now - start

    def prn(self) -> None:
        width = max(len(s.name) for s in self._all) + 1
        for stat in self._all:
            print(f"{stat.name + ':':<{width}} {stat}")

    def brief_prn(self) -> None:
        if self._phase_stack:
            print(f"terminated in: {self._phase_stack[-1][0].name}")
        timed = [
            s
            for s in self._all
            if s.fmt == ".2f" and s not in self._containers and s.value > 0
        ]
        if timed:
            longest = max(timed, key=lambda s: s.value)
            print(f"longest phase: {longest.name} ({longest}s)")
        print(f"{self.its.name}: {self.its}")
        print(f"{self.pures.name}: {self.pures}")


STATS = Stats()


def timed_check(solver, *args):
    """Wrapper for z3 solver check that tracks time in STATS.liatime."""
    STATS.begin_phase(STATS.liatime)
    result = solver.check(*args)
    STATS.end_phase()
    STATS.liacalls += 1
    return result

File: src/qfn2l_cpp/stats.cpp
#include "stats.h"
#include <cstdio>

Stats STATS;

void Stats::begin_phase(TimedStat& s) { phase_stack.push_back({&s, Clock::now()}); }

void Stats::end_phase() {
    auto& ph = phase_stack.back();
    auto elapsed = std::chrono::duration<double>(Clock::now() - ph.start).count();
    ph.stat->value += elapsed;
    phase_stack.pop_back();
}

void Stats::commit_phases() {
    auto now = Clock::now();
    for (auto& ph : phase_stack) {
        auto elapsed = std::chrono::duration<double>(now - ph.start).count();
        ph.stat->value += elapsed;
    }
}

void Stats::prn() const {
    auto prn_count = [](const CountStat& s) {
        std::printf("%-24s %ld\n", s.name, s.value);
    };
    auto prn_timed = [](const TimedStat& s) {
        std::printf("%-24s %.2f\n", s.name, s.value);
    };
    prn_count(its);
    prn_count(pures);
    prn_count(mul_axioms);
    prn_count(mod_axioms);
    prn_count(div_axioms);
    prn_count(liacalls);
    prn_timed(liatime);
    prn_timed(parse_time);
    prn_timed(nnf_time);
    prn_timed(simplify_time);
    prn_timed(propagate_time);
    prn_timed(makedefs_time);
    prn_timed(init_time);
    prn_timed(set_level_time);
    prn_timed(check_nia_time);
    prn_timed(solve_time);
    prn_timed(complete_model_time);
    prn_timed(total_time);
}

void Stats::brief_prn() const {
    if (!phase_stack.empty())
        std::printf("terminated in: %s\n", phase_stack.back().stat->name);

    // Find longest non-container timed stat with value > 0
    std::vector<const TimedStat*> timed = {
        &liatime,        &parse_time,     &nnf_time,
        &simplify_time,  &propagate_time, &makedefs_time,
        &set_level_time, &check_nia_time, &complete_model_time};
    const TimedStat* longest = nullptr;
    for (auto* s : timed) {
        if (s->value > 0 && (!longest || s->value > longest->value))
            longest = s;
    }
    if (longest)
        std::printf("longest phase: %s (%.2fs)\n", longest->name, longest->value);

    std::printf("iteration:  %ld\n", its.value);
    std::printf("pures:      %ld\n", pures.value);
    std::printf("mul_axioms: %ld\n", mul_axioms.value);
    std::printf("div_axioms: %ld\n", div_axioms.value);
    std::printf("mod_axioms: %ld\n", mod_axioms.value);
}
